### PR TITLE
Feat acquisition updates

### LINKED ISSUE #33: Add checksum to file transfers before local deletion
Is your feature request related to a problem? Please describe.
File transfers are currently susceptible to corruption during transfer with no logic checking

Describe the solution you'd like
Add checksum on local vs external file before local deletion

Acceptance Criteria

- [ ] Check sum implemented and unit tested in simulated mode to be working properly.

### LINKED ISSUE #48: Acquisition Name
**Is your feature request related to a problem? Please describe.**
Need a way to have an acquisition name based on metadata 

**Describe the solution you'd like**
Have a way for a user to specify format of acquisition name and then use as a parameter for file saving

**Acceptance Criteria**
- [ ] acquisition name based on metadata
- [ ] updates based on current date and time

### PR DESCRIPTION
**Pull request by adamkglaser**
fixes #48 
fixes #33 

Voxel updates for actual operation...

max projection process updates
- added ability to break up projection in x and y, not just z
- controlled through yaml settings x_projection_px, y_projection_px, z_projection_px
- fixed indexing bug so if the above is not integer multiple of the tile size in xyz, the last projection image is sized correctly
- if projection_px is not defined for xyz, those mips are not generated
- added assertion checks that 0 < projection_px < tile_size_px along xyz
- unit tested and working

file transfer updates
- added file transfer verification using hashing
- checksum would be the best option but is extremely expensive, requires traversing the entire multiple TB file which would take hours
- checking file size alone is simple/fast but would not be robust
- instead implemented partial hashing with https://github.com/kalafut/py-imohash
- imohash checks a user-defined number of bits at the beginning, middle, and end of the file to mostly verify file integrity
- also added timeout and number of retry settings to transfers
- timeout can be triggered if progress monitoring stalls at a given %
- this or other try/catch statements, or failed file hashing would trigger another iteration of file transfers for remaining files. numbe of retry settings controls how many attempts will be made before moving on.
- also unit tested rsync on windows. bug fix was identifying that rsync on windows expects /cygdrive to be appended to the beginning of any path, i.e. /cygdrive/c/users/...
- unit tested and working

adding acquisition name
- we were lacking a mechanism for generating and setting the "acquisition name", i.e. right now exaSPIM_123456_datetime. this has been added as a function within acquisition.py, and is controlled through a name dictionary in the acquisition.yaml.
- example below. delimieter must be specified (defaults to "_"). datetime format, if used, has to be defined and is checked as a valid input to strftime(). format is an ordered list that can call into any keys that exist in the metadata dict, each separated by dict. datetime has to be explicitly listed and does not exist in the metadata dict.

acquisition:
  name:
    delimiter: "\_"
    datetime: "%Y-%m-%d_%H-%M-%S"
    format:
      - instrument_type
      - subject_id
      - datetime
  metadata:
    instrument_type: exaSPIM
    subject_id: 615243
    experimenter_name: adam glaser
    immersion_medium: 0.05X SSC
    immersion_medium_refractive_index: 1.33
    x_anatomical_direction: Anterior_to_posterior
    y_anatomical_direction: Inferior_to_superior
    z_anatomical_direction: Left_to_right

adding acquisition name property to relevant voxel operations
- writers, processes, etc. still need a "path" that is user specified for where to save files. it isn't feasible for the user to define the whole path including the acquisition_name above since the datetime is generated at runtime.
- paths were kept as they are now, and a new property acquisition_name was added to these operations, which is calculated at runtime within the acquisition, and is concatenated with the user specified paths.

### CODE REVIEW

File: voxel/acquisition/acquisition.py [hunk 1]
```diff
@@ -31,6 +32,10 @@ def __init__(self, instrument: Instrument, config_filename: str):
         for device_name, operation_dictionary in self.config['acquisition']['operations'].items():
             self._construct_operations(device_name, operation_dictionary)
 
+    @property
```
Comment (micahwoodard) [inline voxel/acquisition/acquisition.py:35]: We should probably have self._name be initialized as None in case it's called before it's assigned

Comment (adamkglaser) [reply voxel/acquisition/acquisition.py:35]: Thanks good catch. The transfer/MIP changes are probably fine. But any other comments on the "name" implementation? This is the best way I could think to add it, but coded this while quite sleepy and waiting for baby to go back to sleep in the early morning...

**Comment by micahwoodard** [comment]
Looks good to me. We also need to add saving the acquisition yaml file at the end of the scan too or at least the tile list.

File: voxel/acquisition/acquisition.py [hunk 2]
```diff
@@ -69,21 +77,85 @@ def _construct_operations(self, device_name, operation_dictionary):
...
+                        except:
+                            raise ValueError(f'{datetime_format} is not a valid format for strftime.')
+                elif name not in metadata_dict.keys():
+                    raise ValueError(f'{name} is not a valid metadata property. check yaml file.')
+                else:
+                    metadata_value = metadata_dict[name]
+                    self._name += str(metadata_value)
+                if name != name_dict['format'][-1]:
+                    self._name += delimiter
+
```
Comment (micahwoodard) [inline voxel/acquisition/acquisition.py:119]: I think this is an acceptable way of doing this. My only qualm (and I have the same thing in gui_yaml) is that if we're going to have dictionaries/list in the yaml that contain unclear keys like format = ['instrument_type', 'subject_id', 'datetime'], I feel like we need some way to validate or inform the user of the options of those keywords.  We can even create a custom AcquisitionName class that has the options outlined within the class and then you can create a clear gui or a user can go into the class to see the options. Do you have other ideas on how to handle informing the users what keywords the acquisition is looking for?

Comment (adamkglaser) [reply voxel/acquisition/acquisition.py:119]: I totally agree. This crossed my mind when implementing. In this case, I was thinking that the only enforcement that is needed is that the "format" spec only references keys that actually exist in the YAML file. In this way the user can put anything they want in there, and then reference to it for creating any filename format spec they want. I think right now the metadata dictionary does not get used anywhere else in the code? So nothing is dependent on it, it just gets used for metadata purposes and now - after this addition - how the folder/files get named. 

In our case, we will use keys in metadata to generate the schema JSONs, as well as name our datasets in the proper way. Others could adapt as needed, there would be no need to restrict what options are available?

File: voxel/acquisition/acquisition.py [hunk 3]
```diff
@@ -31,6 +32,10 @@ def __init__(self, instrument: Instrument, config_filename: str):
         for device_name, operation_dictionary in self.config['acquisition']['operations'].items():
             self._construct_operations(device_name, operation_dictionary)
 
+    @property
+    def name(self):
```
Comment (micahwoodard) [inline voxel/acquisition/acquisition.py:36]: Can we add a docstring here describing what name means in this context?

File: voxel/acquisition/acquisition.py [hunk 4]
```diff
@@ -69,21 +77,85 @@ def _construct_operations(self, device_name, operation_dictionary):
                 getattr(self, operation_type)[device_name] = {}
             getattr(self, operation_type)[device_name][operation_name] = operation_object
 
+    def _construct_acquisition_name(self):
+        if not self.acquisition['name']:
```
Comment (micahwoodard) [inline voxel/acquisition/acquisition.py:81]: Can we have a docstring here to describe what this function is  doing?

File: voxel/acquisition/acquisition.py (hunk 2 above)
Comment (micahwoodard) [reply voxel/acquisition/acquisition.py:119]: I think we should probably create a metadata class that is akin to a device class that can handle this internally. So something like 

```
from datetime import datetime
import logging

DATETIME_FORMATS = {'year/month/day/hour/minute/second': '%Y-%m-%d_%H-%M-%S',
                    'month/day/year/hour/minute/second': '%m-%d-%Y_%H-%M-%S',
                    'month/day/year': '%m-%d-%Y',
                    'None': None}

class MetadataClass:

    def __init__(self, metadata_dictionary: dict, date_format: str = 'None', name_specs: dict = {}):

        self.log = logging.getLogger(f"{__name__}.{self.__class__.__name__}")

        for key, value in metadata_dictionary.items():
            # create properties from keyword entries
            setattr(self, f'_{key}', value)
            new_property = property(fget=lambda x, k=key: getattr(self, f'_{k}'),
                                    fset=lambda val, k=key: setattr(self, f'_{k}', val))
            setattr(type(self), key, new_property)
        self.date_format = date_format
        self._acquisition_name = self.generate_acquisition_name(name_specs)

    @property
    def date_format(self):
        return self._date_format

    @date_format.setter
    def date_format(self, format):
        if format not in list(DATETIME_FORMATS.keys()):
            raise ValueError(f'{format} is not a valid datime format. Please choose from {DATETIME_FORMATS.keys()}')
        self._date_format = DATETIME_FORMATS[format]

    @property
    def acquisition_name(self):
        """Unique name that descibes acquisition adhering to passed in name_specs and date of acquisition"""
        return self._acquisition_name

    def generate_acquisition_name(self, name_specs: dict):
        """Function to generate name of acquisition based on format"""

        if name_specs == {}:
            return 'test'

        else:
            # grab the delimeter
            if 'delimiter' not in name_specs.keys():
                self.log.warning('no delimiter specified in yaml file. defaulting to "_".')
            delimiter = name_specs.get('delimiter', '_')

            # grab the name format list
            format = name_specs.get('format', [])
            name = []
            for prop_name in format:
                if not isinstance(getattr(type(self), prop_name, None), property):  # check if prop name is metadata property
                    raise ValueError(f'{prop_name} is not a metadata property. Please choose from {self.__dir__()}')
                name.append(str(getattr(self, prop_name)))
            date = datetime.now().strftime(self._date_format) if self._date_format is not None else ''
            name.append(date)
            return f'{delimiter}'.join(name)
```

And then using it looks something like this: 

```
metadata_dictionary = {
        'instrument_type': 'exaspim',
        'subject_id': 123456,
        'experimenter_name': 'adam glaser',
        'immersion_medium': '0.05XSSC',
        'immersion_medium_refractive_index': 1.33,
        'x_anatomical_direction': 'Anterior_to_posterior',
        'y_anatomical_direction': 'Inferior_to_superior',
        'z_anatomical_direction': 'Left_to_right'}

    datetime_format = 'year/month/day/hour/minute/second'

    name_specs = {
        'deliminator': '_',
        'format': ['instrument_type', 'subject_id']}

    metadata_class = MetadataClass(metadata_dictionary, datetime_format, name_specs)
    <exaspim_123456_2024-07-11_15-23-53>

```

Comment (adamkglaser) [reply voxel/acquisition/acquisition.py:119]: I am on board with this. Thanks for taking a crack at this!

**Comment by micahwoodard** [comment]
@adamkglaser I'm getting bugs when trying to use max_projection.py. Did you run into this problem? 
image

**Comment by adamkglaser** [comment]
> @adamkglaser I'm getting bugs when trying to use max_projection.py. Did you run into this problem? Should _z_projection_count_px initialize to 0 instead of none to avoid this problem? image

Hmm it looks like a bug because z_projection_count_px isn't defined. If you define it does the code work? I might not have checked the case where the count in z is undefined. We should default it to equal frame_count_px if not defined? Or simply ignore doing the MIP in z, I believe if X and Y are not defined they are ignored.

**Comment by micahwoodard** [comment]
@adamkglaser I think you accounted if self._z_projection_count_px == None [earlier ](https://github.com/AllenNeuralDynamics/voxel/blob/374b875fbd682033cd46dded590404c9ebd81d0c/voxel/processes/cpu/max_projection.py#L166C9-L172C103)but then assume self._z_projection_count_px won't be None later in the [while frame_index < self._frame_count_px_px loop](https://github.com/AllenNeuralDynamics/voxel/blob/374b875fbd682033cd46dded590404c9ebd81d0c/voxel/processes/cpu/max_projection.py#L177C9-L178C68). This either needs to be fixed or the self._z_projection_count_px should be initialized to be something that makes more sense

**Comment by adamkglaser** [comment]
> @adamkglaser I think you accounted if self._z_projection_count_px == None [earlier ](https://github.com/AllenNeuralDynamics/voxel/blob/374b875fbd682033cd46dded590404c9ebd81d0c/voxel/processes/cpu/max_projection.py#L166C9-L172C103)but then assume self._z_projection_count_px won't be None later in the [while frame_index < self._frame_count_px_px loop](https://github.com/AllenNeuralDynamics/voxel/blob/374b875fbd682033cd46dded590404c9ebd81d0c/voxel/processes/cpu/max_projection.py#L177C9-L178C68). This either needs to be fixed or the self._z_projection_count_px should be initialized to be something that makes more sense

this is now fixed. i think this PR is now OK to merge.

**Comment by micahwoodard** [comment]
Yay thanks!

**Comment by adamkglaser** [comment]
@micahwoodard this one is OK to merge now?

**Pull request merged**